src/ebay-listbox/listbox.tsx

```diff
@@ -0,0 +1,217 @@
+import { createLinear } from 'makeup-active-descendant'
+import typeahead from 'makeup-typeahead'
+import classNames from 'classnames'
+import { ComponentProps, ReactNode, useRef, useState } from 'react'
+import { filterByType, scroll } from '../utils'
+import { EbayListboxOption } from './listbox-option'
+import { EbayChangeEventHandler } from '../events'
+
+const TYPEAHEAD_TIMEOUT_LENGTH = 1300
+
+export type ChangeEventProps = {
+    index: number;
+    selected: string[];
+    wasClicked: boolean;
+}
+
+export type EbayListboxProps = Omit<ComponentProps<'div'>, 'onChange'> & {
+    name?: string;
+    disabled?: boolean;
+    children: ReactNode;
+    listSelection?: 'auto' | 'manual';
+    typeaheadTimeoutLength?: number;
+    maxHeight?: string | number;
+    selectClassName?: string;
+    onChange?: EbayChangeEventHandler<HTMLSpanElement, ChangeEventProps>;
+    onEscape?: () => void
+};
+
+export const EbayListbox: FC<EbayListboxProps> = ({
+    name,
+    className,
+    disabled,
+    children,
+    tabIndex = 0,
+    listSelection,
+    maxHeight,
+    typeaheadTimeoutLength,
+    selectClassName,
+    onChange = () => {},
+    onEscape = () => {},
+    ...rest
+}) => {
+    const options = filterByType(children, EbayListboxOption)
+    const selectedOptionComponentIndex = options.findIndex(option => option.props.selected)
+
+    const [selectedIndex, setSelectedIndex] = useState(
+        options.findIndex(option => option.props.selected)
+    )
+
+    useEffect(() => {
+        if (selectedOptionComponentIndex !== selectedIndex) {
+            setSelectedIndex(selectedOptionComponentIndex)
+        }
+    }, [selectedOptionComponentIndex])
+
+    const containerRef = useRef<HTMLDivElement>()
+    const activeDescendantRef = useRef<ReturnType<typeof createLinear>>()
+    const typeaheadRef = useRef<ReturnType<typeof typeahead>>()
+    const wasClickedRef = useRef(false)
+
+    function handleChange(event: CustomEvent | Event | KeyboardEvent, nextSelectedIndex: number, wasClicked: boolean) {
+        if (nextSelectedIndex === selectedIndex) {
+            return
+        }
+
+        const option = options[nextSelectedIndex]
+
+        if (option.props.disabled) {
+            return
+        }
+
+        setSelectedIndex(nextSelectedIndex)
+        onChange(event as unknown as ChangeEvent<HTMLSpanElement>, {
+            index: nextSelectedIndex,
+            selected: [options[nextSelectedIndex].props.value],
+            wasClicked
+        })
+    }
+
+    function handleKeyDown(event: KeyboardEvent) {
+        switch (event.code) {
+            case 'Esc':
+            case 'Escape':
+                onEscape()
+                break
+            case 'Space':
+            case 'Enter':
+                handleChange(event, activeDescendantRef.current.index, false)
+                break
+            default:
+                break
+        }
+
+        const itemIndex = typeaheadRef.current.getIndex(
```

Comment:
should we do null check after the ref exists to avoid errors -- or is it safe 100%?